lib/ukfs-virtiofs/virtiofs-fuse.c

```diff
@@ -0,0 +1,1057 @@
+		{ &ohdr, sizeof(ohdr) }
+	};
+	ssize_t rlen;
+
+	init_hdr(fdev, &hdr, FUSE_RELEASE, sizeof(in));
+	rlen = uk_virtiofs_request(fdev->dev, iov, 2, 1);
+	return check_reply(rlen, &ohdr);
+}
+
+int virtiofs_fuse_fsync(struct virtiofs_fuse_dev *fdev,
+			uint64_t node, uint64_t fh, uint32_t fsync_flags)
+{
+	struct fuse_in_header hdr = {
+		.nodeid = node
+	};
+	struct fuse_fsync_in in = {
+		.fh = fh,
+		.fsync_flags = fsync_flags
+	};
+	struct fuse_out_header ohdr;
+	const struct uk_virtiofs_iovec iov[] = {
+		{ &hdr, sizeof(hdr) },
+		{ &in, sizeof(in) },
+		{ &ohdr, sizeof(ohdr) }
+	};
+	ssize_t rlen;
+
+	init_hdr(fdev, &hdr, FUSE_FSYNC, sizeof(in));
+	rlen = uk_virtiofs_request(fdev->dev, iov, 2, 1);
+	return check_reply(rlen, &ohdr);
+}
+
+/* No xattr support */
+
+int virtiofs_fuse_flush(struct virtiofs_fuse_dev *fdev,
+			uint64_t node, uint64_t fh)
+{
+	struct fuse_in_header hdr = {
+		.nodeid = node
+	};
+	struct fuse_flush_in in = {
+		.fh = fh
+	};
+	struct fuse_out_header ohdr;
+	const struct uk_virtiofs_iovec iov[] = {
+		{ &hdr, sizeof(hdr) },
+		{ &in, sizeof(in) },
+		{ &ohdr, sizeof(ohdr) }
+	};
+	ssize_t rlen;
+
+	init_hdr(fdev, &hdr, FUSE_FLUSH, sizeof(in));
+	rlen = uk_virtiofs_request(fdev->dev, iov, 2, 1);
+	return check_reply(rlen, &ohdr);
+}
+
+ssize_t virtiofs_fuse_readdir(struct virtiofs_fuse_dev *fdev, uint64_t node,
+			      uint64_t fh, uint64_t off,
+			      void *buf, uint64_t len)
+{
+	struct fuse_in_header hdr = {
+		.nodeid = node
+	};
+	struct fuse_read_in in = {
+		.fh = fh,
+		.offset = off,
+		.size = len
+	};
+	struct fuse_out_header ohdr;
+	const struct uk_virtiofs_iovec iov[] = {
+		{ &hdr, sizeof(hdr) },
+		{ &in, sizeof(in) },
+		{ &ohdr, sizeof(ohdr) },
+		{ buf, len }
+	};
+	ssize_t rlen;
+	int err;
+
+	init_hdr(fdev, &hdr, FUSE_READDIR, sizeof(in));
+	rlen = uk_virtiofs_request(fdev->dev, iov, 2, 2);
+	if (unlikely((err = check_reply(rlen, &ohdr))))
+		return err;
+	return payload_len(rlen);
+}
+
+/* No lock ops support */
+
+int virtiofs_fuse_access(struct virtiofs_fuse_dev *fdev,
+			 uint64_t node, uint32_t mask)
+{
+	struct fuse_in_header hdr = {
+		.nodeid = node
+	};
+	struct fuse_access_in in = {
+		.mask = mask
+	};
+	struct fuse_out_header ohdr;
+	const struct uk_virtiofs_iovec iov[] = {
+		{ &hdr, sizeof(hdr) },
+		{ &in, sizeof(in) },
+		{ &ohdr, sizeof(ohdr) }
+	};
+	ssize_t rlen;
+
+	init_hdr(fdev, &hdr, FUSE_ACCESS, sizeof(in));
+	rlen = uk_virtiofs_request(fdev->dev, iov, 2, 1);
+	return check_reply(rlen, &ohdr);
+}
+
+int virtiofs_fuse_create(struct virtiofs_fuse_dev *fdev,
+			 uint64_t node, const char *name, size_t len,
+			 const struct fuse_create_in *in,
+			 union virtiofs_fuse_entry_outp entry,
+			 struct fuse_attr *entry_attr,
+			 struct fuse_open_out *out)
+{
+	struct fuse_in_header hdr = {
+		.nodeid = node
+	};
+	struct fuse_out_header ohdr;
+	struct uk_virtiofs_iovec iov[] = {
+		{ &hdr, sizeof(hdr) },
+		{ (void *)in, sizeof(*in) },
+		{ (void *)name, len + 1 },
```

Comment:
Why does this one has `len + 1` (emphasis on `+1`) but the `virtiofs_fuse_lookup` doesn't?